code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go

```diff
@@ -0,0 +1,162 @@
+
+	Fee           currency.Coin `json:"fee"`
+	UnstakeTotal  currency.Coin `json:"unstake_total"`
+	TotalStake    currency.Coin `json:"total_stake"`
+	ServiceCharge float64       `json:"service_charge"`
+}
+
+func (edb *EventDb) ReplicateAuthorizerAggregate(p common.Pagination) ([]AuthorizerAggregate, error) {
+	var snapshots []AuthorizerAggregate
+
+	queryBuilder := edb.Store.Get().
+		Model(&AuthorizerAggregate{}).Offset(p.Offset).Limit(p.Limit)
+
+	queryBuilder.Order(clause.OrderByColumn{
+		Column: clause.Column{Name: "id"},
+		Desc:   false,
+	})
+
+	result := queryBuilder.Scan(&snapshots)
+	if result.Error != nil {
+		return nil, result.Error
+	}
+
+	return snapshots, nil
+}
+
+// nolint
+func (edb *EventDb) updateAuthorizerAggregate(round, pageAmount int64, gs *globalSnapshot) {
+	count, err := edb.GetAuthorizerCount()
+	if err != nil {
+		logging.Logger.Error("update_authorizer_aggregates", zap.Error(err))
+		return
+	}
+	size, currentPageNumber, subpageCount := paginate(round, pageAmount, count, edb.PageLimit())
+
+	exec := edb.Store.Get().Exec("CREATE TEMP TABLE IF NOT EXISTS temp_ids "+
+		"ON COMMIT DROP AS SELECT id as id FROM authorizers ORDER BY (id, creation_round) LIMIT ? OFFSET ?",
+		size, size*currentPageNumber)
+	if exec.Error != nil {
+		logging.Logger.Error("error creating temp table", zap.Error(exec.Error))
+		return
+	}
+
+	for i := 0; i < subpageCount; i++ {
+		edb.calculateAuthorizerAggregate(gs, round, edb.PageLimit(), int64(i)*edb.PageLimit())
+	}
+}
+
+// nolint
+func (edb *EventDb) calculateAuthorizerAggregate(gs *globalSnapshot, round, limit, offset int64) {
+
+	var ids []string
+	r := edb.Store.Get().
+		Raw("select id from temp_ids ORDER BY ID limit ? offset ?", limit, offset).Scan(&ids)
+	if r.Error != nil {
+		logging.Logger.Error("getting ids", zap.Error(r.Error))
+		return
+	}
+	logging.Logger.Debug("getting ids", zap.Strings("ids", ids))
+
+	var currentAuthorizers []Authorizer
+	result := edb.Store.Get().
+		Raw("SELECT * FROM Authorizers WHERE id in (select id from temp_ids ORDER BY ID limit ? offset ?)", limit, offset).
+		Scan(&currentAuthorizers)
+	if result.Error != nil {
+		logging.Logger.Error("getting current Authorizers", zap.Error(result.Error))
+		return
+	}
+	logging.Logger.Debug("Authorizer_snapshot", zap.Int("total_current_Authorizers", len(currentAuthorizers)))
+
+	if round <= edb.AggregatePeriod() && len(currentAuthorizers) > 0 {
+		if err := edb.addAuthorizerSnapshot(currentAuthorizers); err != nil {
+			logging.Logger.Error("saving Authorizers snapshots", zap.Error(err))
+		}
+	}
+
+	oldAuthorizers, err := edb.getAuthorizerSnapshots(limit, offset)
+	if err != nil {
+		logging.Logger.Error("getting Authorizer snapshots", zap.Error(err))
+		return
+	}
+	logging.Logger.Debug("Authorizer_snapshot", zap.Int("total_old_Authorizers", len(oldAuthorizers)))
+
+	var aggregates []AuthorizerAggregate
+	for _, current := range currentAuthorizers {
```

Comment:
use common code to process Provider fields, avoid copy paste

Reply:
The for loop is similar but the fields vary, so how to use the `common code`.

Reply:
actually fields are the same, but on different structures, we can abstract them out by the interface for example, 

```go
type ProviderAggregate Interface {
    SetTotalStake(newVal int64)
    SetUnstakeTotal(newVal int64)
    ...
    TotalStake() int64
    TotalUnstake() int64
}
```
then we can common code that takes 2 snapshots and assigns one

```go
recalculateProviderFields(new, old ProviderAggregates) {
    new.SetTotalStake(old.TotalStake())
  ....
}

```

Reply:
UPD: we can do it better with generics, if you need to return new aggregate from recalculate

```go
recalculateProviderFields[T ProviderAggregates](new, old ProviderAggregates) T {
    new.SetTotalStake(old.TotalStake())
  ....
    return newAggregate
}
```